3단계 - 로또(자동)

Ttungga opened the pull request:
안녕하세요 지민님.

3단계 미션 진행 후 PR 생성하였습니다. 리뷰 부탁드립니다.

TDD로 최대한 진행해보았는데 아직 어색하네요..ㅠㅠ

이번 미션을 수행하면서 객체의 책임에 대해서 고민을 많이 하게 되었는데요. 이에 따른 몇가지 질문이 있어서 아래에 정리하였습니다.
확인 후 답변 부탁드리겠습니다.

1. LottoResult라는 클래스가 UI 로직을 가지게 구현하였는데요. 도메인 객체가 UI 로직을 가져도 될지 궁금합니다.
2. 1번 질문과 관련하여 LottoResult 객체에서 결과 계산의 책임만 갖도록 하고 LottoPresenter 객체만 UI 로직을 갖도록 리팩토링할지에 대해 고민이 되었습니다. 이렇게 변경한다면  이미 LottoTickets와 LottoPayment를 가진 클래스라면 누구나 결과를 계산할 수 있는데 굳이 LottoResult 객체가 존재할 필요가 있을까? 하는 생각이 들더라구요..
3. validation 메서드가 반환값을 가져도 괜찮은지 궁금합니다.
4. 일급컬렉션에서 스태틱 멤버를 가져도 될까요?

Ttungga committed:
refactor: remove SetTest.convertIntegerToBoolean

Ttungga committed:
refactor: add static variables for matched groups' indexes of input with a custom separator

Ttungga committed:
refactor: create calculator package and move files

Ttungga committed:
docs: add a function list to README.md

Ttungga committed:
feat: add Payment

Ttungga committed:
feat: add validation logics to Payment

- add Payment.validateMoney and Payment.checkAmount

Ttungga committed:
docs: add a function for calculating purchasable amount of lotto tickets

Ttungga committed:
feat: add toString, equals, hashCode to Payment

Ttungga committed:
feat: add purchasableAmount field and getter to Payment

Ttungga committed:
feat: add LottoTicket

Ttungga committed:
docs: add a validating function for lotto numbers

Ttungga committed:
feat: add validation logics to LottoTicket

- add LottoTicket.validateNumbers, LottoTicket.checkAmountOfNumbers, and LottoTicket.checkRangeOfNumber

Ttungga committed:
refactor: change field of LottoTicket type array to list and modify tests

Ttungga committed:
feat: add LottoTickets

Ttungga committed:
fix: modify subList index and change not to use subList

Ttungga committed:
docs: add a domain of lotto numbers

- move validating function for lotto numbers from lotto ticket object to lotto numbers object
- remove the domain of winning numbers

Ttungga committed:
feat: add LottoNumbers and modify LottoTicket and LottoTickets

Ttungga committed:
docs: add a domain for auto lotto numbers issuer

Ttungga committed:
feat: add AutoLottoNumbersIssuer and modify LottoTickets to use it

Ttungga committed:
docs: add a function for checking prizes

Ttungga committed:
feat: add logics for checking prizes

- add Prize
- add LottoNumbers.matches and LottoTicket.prize

Ttungga committed:
feat: add LottoResult

Ttungga committed:
docs: remove a function for statistic object

Ttungga committed:
feat: add a function for prizes of lotto tickets

Ttungga committed:
feat: add LottoResult

- add Payment.getMoney
- add Prize.resultMessage
- add LottoResult

Ttungga committed:
refactor: rename Payment to LottoPayment

Ttungga committed:
refactor: change fields of LottoResult

Ttungga committed:
refactor: change fields of LottoTickets

Ttungga committed:
feat: add a constructor of LottoNumbers for string parameter

- add LottoNumbersStringConverter

Ttungga committed:
feat: add a view logic and main method

- add LottoPresenter
- add print logic to LottoTickets, LottoTicket, LottoNumbers

joojimin commented on README.md:
```diff
@@ -1,9 +1,31 @@
+  * 수익률 계산
+
+
+* 입력 및 결과 안내 메시지를 출력한다.
```
요구사항정리 👍 👍 👍 👍 👍

joojimin commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -40,7 +43,7 @@ private static int sumNumbers(final List<Integer> numbers) {
         Matcher matcher = CUSTOM_SEPARATOR_PATTERN.matcher(input);
         if (matcher.find()) {
-            return matcher.group(2).split(matcher.group(1));
+            return matcher.group(NUMBERS_GROUP_INDEX).split(matcher.group(CUSTOM_SEPARATOR_GROUP_INDEX));
```
피드백 반영 👍 👍

joojimin commented on src/main/java/lotto/domain/LottoNumbers.java:
```diff
@@ -0,0 +1,76 @@
+public class LottoNumbers {
+    public static final int AMOUNT_OF_NUMBERS = 6;
+    public static final int MINIMUM_NUMBER = 1;
+    public static final int MAXIMUM_NUMBER = 45;
```
4. 일급컬렉션에서 스태틱 멤버를 가져도 될까요?
=> 해당질문이 이부분인것 같은데요!
일급컬렉션이 스태틱멤버를 갖지말아야한다는 조건이나 코드규칙은 못본것같습니다! 다만 위처럼 같은 성격의 static 멤버가 많다면 Enum으로 따로 분리해서 분리하는 것도 좋을거같아요!

joojimin commented on src/main/java/lotto/domain/LottoNumbers.java:
```diff
@@ -0,0 +1,76 @@
+
+    public LottoNumbers(final String numbersString) {
+        this(LottoNumbersStringConverter.convert(numbersString));
+    }
```
팩토리메서드 생성자를 이용하여 내부에서하는 convert 행위까지 네이밍으로 알려주는건 어떨까요? 파라미터만 보고는 정확한 행위를 파악하기힘들것같습니다

joojimin commented on src/main/java/lotto/domain/LottoPayment.java:
```diff
@@ -0,0 +1,60 @@
+
+import java.util.Objects;
+
+public class LottoPayment {
```
로또 금액 분리 👍 👍

joojimin commented on src/main/java/lotto/domain/LottoPayment.java:
```diff
@@ -0,0 +1,60 @@
+
+    private static void checkAmount(final int money) {
+        if (money <= 0 || money % LOTTO_TICKET_COST != 0) {
+            throw new IllegalArgumentException();
```
Exception 메시지도 함께 적어주면 좋을것 같습니다

joojimin commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.Objects;
+
+public class LottoResult {
+    private static final String RESULT_ANNOUNCEMENT_MESSAGE = "\r\n당첨 통계\r\n---------";
```
OS마다 줄바꿈이 다르기때문에 `System.lineSeparator()` 같은 코드를 사용하는것이 좋을거같습니다!

joojimin commented on src/test/java/SetTest.java:
```diff
@@ -38,17 +38,10 @@ void when_call_contains_for_elements_in_Set_should_return_true(final int element
+    @CsvSource(value = {"1:true", "2:true", "3:true", "4:false", "5:false"}, delimiter = ':')
     @DisplayName("1, 2, 3이 들어있는 Set 객체에 contains 메서드를 호출하였을 때 파라미터에 따라 올바른 결과가 반환되어야 한다")
-    void when_call_contains_for_elements_in_Set_should_return_correct_result(final int input, final int expected) {
+    void when_call_contains_for_elements_in_Set_should_return_correct_result(final int input, final boolean expected) {
```
피드백 반영 👍 👍

joojimin commented on src/test/java/lotto/domain/LottoNumbersTest.java:
```diff
@@ -0,0 +1,153 @@
+        numbersList.add(greaterThanSix);
+
+        // when and then
+        for (List<Integer> numbers : numbersList) {
```
직접 만든 List로 반복하여 assertThatThrownBy을 돌리면 오류가나는 시점에서 끊길것같아요! 
`@parameterizedtest` 를 활용해보시는 것도 좋을것같네요!

joojimin commented on src/test/java/lotto/domain/LottoNumbersTest.java:
```diff
@@ -0,0 +1,153 @@
+        final LottoNumbers twoMatchesNumbers = new LottoNumbers(Arrays.asList(2, 5, 34, 37, 41, 44));
+
+        // when and then
+        assertThat(noMatchesNumbers.matches(winningNumbers)).isEqualTo(0);
```
반복되는 테스트에 주어진 given 조건만 다르다면 앞서 드린 피드백과 같은이유로 @parameterizedtest 를 활용하면 좋을 것 같습니다!

joojimin reviewed (changes requested):
안녕하세요~ 민철님!
주지민입니다!

다음미션 가기전에 생각해보면 좋을거같은 피드백 몇가지 드렸습니다!
확인해보시고 다시 리뷰요청부탁드려요~!

1. LottoResult라는 클래스가 UI 로직을 가지게 구현하였는데요. 도메인 객체가 UI 로직을 가져도 될지 궁금합니다.
=> LottoResult 객체를 보면 결과 계산의 책임외에도 UI 역할을 담당하는 부분이 강하게 결합되어있어서 테스트하기도 힘들거같아요! 아래 2번질문에서 생각하신 것과 동일하게 저도생각합니다~

2. 1번 질문과 관련하여 LottoResult 객체에서 결과 계산의 책임만 갖도록 하고 LottoPresenter 객체만 UI 로직을 갖도록 리팩토링할지에 대해 고민이 되었습니다. 이렇게 변경한다면 이미 LottoTickets와 LottoPayment를 가진 클래스라면 누구나 결과를 계산할 수 있는데 굳이 LottoResult 객체가 존재할 필요가 있을까? 하는 생각이 들더라구요..
=> 말씀하신대로 각 클래스가 하나의 책임만 갖도록 설계하셧는데 LottoResult 객체가 불필요하다면 과감히 지우시는 것도 좋을것같습니다 ㅎㅎ



3. validation 메서드가 반환값을 가져도 괜찮은지 궁금합니다.
=> 반환값을 가지면 안된다는 규약을 본적은 없는것같아요! 다만 validation이라고 하면 타겟이되는 대상을 검증한다, 즉 검증이 맞다, 아니다의 결과가 나오길 기대하는 부분이 많으니 해당 메서드에 검증하는 책임 + 다른 책임이 더해져 반환값을 만든다면 책임소개가 뒤섞일거같습니다!


4. 일급컬렉션에서 스태틱 멤버를 가져도 될까요?
=> 리뷰로 말씀드렸습니다!

Ttungga committed:
refactor: change @Test to @ParameterizedTest

Ttungga committed:
refactor: remove unnecessary statement

Ttungga committed:
refactor: extract messages to enums

Ttungga committed:
refactor: add LottoNumbersCondition

Ttungga committed:
refactor: add message package

Ttungga committed:
refactor: rollback rearrangement

Ttungga committed:
refactor: move responsibilities to validate and create LottoNumbers to LottoNumbersFactory

- move validation logics of LottoNumbers to LottoNumbersFactory
- remove LottoNumbers constructor having String parameter

Ttungga committed:
refactor: move responsibilities to create LottoTickets automatically to LottoTicketsFactory

Ttungga committed:
refactor: move responsibilities to validate and create LottoPayment to LottoPaymentFactory

Ttungga committed:
refactor: change domain classes' access modifiers to default

Ttungga committed:
fix: change AutoNumbersIssuer.issueNumbers to return List<Integer>

- rename AutoLottoNumbersIssuer to AutoNumbersIssuer

Ttungga committed:
refactor: remove LottoResult and add PrizeCalculator

Ttungga commented:
안녕하세요 지민님.
먼저 상세히 리뷰해주셔서 정말 감사드립니다!!

말씀해주신 부분들을 수정해보았는데요. 다시 한번 리뷰 부탁드리겠습니다 _ _

대부분의 피드백은 그대로 수정하였는데 LottoNumbers에 팩토리 메서드를 추가하려다가,
모든 주요 도메인들의 생성자 접근제어자를 변경하고 각각 팩토리 클래스 생성하는 것으로 방향을 바꾸게 되었습니다.
도메인의 생성과 이에 필요한 유효성 검증에 대한 책임을 굳이 도메인 객체가 가져야 하나 하는 생각이 들어서요.
이렇게 책임을 분배하는 것도 적절할까요?

또한 몇가지 궁금한 부분이 생겨서 아래에 정리해보았는데요. 확인 후 답변 부탁드리겠습니다.
1. (실제로 사용하지는 않았지만) 정적 팩토리 메서드를 사용하는 클래스에서는 모든 생성자를 정적 팩토리 메서드로 변경해야할지 궁금합니다.
2. 클래스 구조에서 private static 메서드의 적절한 위치기 궁금합니다. public static 메서드 다음에 위치하는 것과 public instance 메서드 다음에 위치하는 것 중에 어떤 게 좋을지 판단이 잘 안 되더라구요..

joojimin commented on src/test/java/lotto/domain/LottoTicketsTest.java:
```diff
@@ -0,0 +1,54 @@
+import org.junit.jupiter.api.Test;
+
+class LottoTicketsTest {
+    private final LottoNumbers lottoNumbers = new LottoNumbers(Arrays.asList(1, 2, 3, 4, 5, 6));
```
테스트전 초기화할 객체가 필요하다면 JUnit의 `@Before` 어노테이션을 이용해서 각 테스트마다 초기화시키는게 좋을것같아요! 혹시나모를 테스트간 충돌을 방지할수있습니다!

joojimin commented on src/test/java/calculator/StringAddCalculatorTest.java:
```diff
@@ -7,7 +9,7 @@
     @Test
     @DisplayName("입력값이 null 또는 빈 문자열이면 0이 반환되어야 한다")
-    public void splitAndSum_null_또는_빈문자() {
+    void splitAndSum_null_또는_빈문자() {
```
불필요한 접근지정자, throws 처리도 지워주셧네요!! ㅎㅎ 👍 👍 👍

joojimin reviewed (approved):
안녕하세요~ 민철님!!
피드백 고민하고 수정하시느냐고 고생많으셧습니다!
이번미션은 여기서 머지하고 다음미션에서 같이 고민해봐요!! 화이팅입니다!


대부분의 피드백은 그대로 수정하였는데 LottoNumbers에 팩토리 메서드를 추가하려다가,
모든 주요 도메인들의 생성자 접근제어자를 변경하고 각각 팩토리 클래스 생성하는 것으로 방향을 바꾸게 되었습니다.
도메인의 생성과 이에 필요한 유효성 검증에 대한 책임을 굳이 도메인 객체가 가져야 하나 하는 생각이 들어서요.
이렇게 책임을 분배하는 것도 적절할까요?
=> 유효성 검증이 도메인의 생성과 같이하기에 너무 거대하다면 이또한 클래스를 분리시켜보는건 어떨까요?
저도 주로 간단한 파라미터 검증만 도메인 생성부분에서 검증하고 비즈니스로직과 결합하여 유효성을 검증할땐 validation의 역할을 하는 클래스를 따로 만드는편인데요! 이렇게하면 책임소재가 명확해져서 유효성부분이 변경되거나 확장될때 유용했습니다!


(실제로 사용하지는 않았지만) 정적 팩토리 메서드를 사용하는 클래스에서는 모든 생성자를 정적 팩토리 메서드로 변경해야할지 궁금합니다.
=> 무조건적으로 정적팩토리메서드로 변경하는것은 좋지않을것같아요! 생성자에 다른 로직이 들어가거나 네이밍을 줄수있다는 장점으로 정적팩토리메서드를 사용하곤하지만 불필요한곳에 지나치게 남발하면 오히려 혼란을 줄수있을 것같습니다!


클래스 구조에서 private static 메서드의 적절한 위치기 궁금합니다. public static 메서드 다음에 위치하는 것과 public instance 메서드 다음에 위치하는 것 중에 어떤 게 좋을지 판단이 잘 안 되더라구요..
=> 일반적으로 static(클래스) 메서드들은 상단으로 올리는 경향이 있는데요! 클린코드에서는 public static -> public instance 메서드 순으로 작성하고 private static, private 메서드를 호출되는 부분과 가깝게 두는것이 좋다라고 제시하고있네요! 이렇게 작성하면 클래스가 커짐에 따라 메서드를 찾는 거리가 좁혀져서 가독성에 도움된다고 본것같습니다!
